platform/api.scripting/src/org/netbeans/modules/scripting/GraalEngine.java

```
@@ -0,0 +1,175 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.scripting;
+
+import java.io.Reader;
+import javax.script.Bindings;
+import javax.script.Invocable;
+import javax.script.ScriptContext;
+import javax.script.ScriptEngine;
+import javax.script.ScriptException;
+import org.graalvm.polyglot.PolyglotException;
+import org.graalvm.polyglot.Source;
+import org.graalvm.polyglot.Value;
+
+final class GraalEngine implements ScriptEngine, Invocable {
+
+    private final GraalEngineFactory factory;
+
+    GraalEngine(GraalEngineFactory f) {
+        this.factory = f;
+    }
+
+    private String id() {
+        return factory.id;
+    }
+
+    @Override
+    public GraalContext getContext() {
+        return factory.ctx;
+    }
+
+    @Override
+    public GraalEngineFactory getFactory() {
+        return factory;
+    }
+
+    @Override
+    public Object eval(String src, ScriptContext arg1) throws ScriptException {
+        Value result = evalImpl(arg1, src);
+        return unbox(result);
+    }
+
+    private Value evalImpl(ScriptContext arg1, String src) throws ScriptException {
+
+        try {
+            return ((GraalContext) arg1).ctx().eval(id(), src);
+        } catch (PolyglotException e) {
+            throw new ScriptException(e);
+        }
+    }
+
+    @Override
+    public Object eval(Reader arg0, ScriptContext arg1) throws ScriptException {
+        Source src = Source.newBuilder(id(), arg0, null).buildLiteral();
+        Value result = ((GraalContext)arg1).ctx().eval(src);
+        return unbox(result);
+    }
+
+    @Override
+    public Object eval(String arg0) throws ScriptException {
+        return eval(arg0, factory.ctx);
+    }
+
+    @Override
+    public Object eval(Reader arg0) throws ScriptException {
+        return eval(arg0, factory.ctx);
+    }
+
+    @Override
+    public Object eval(String arg0, Bindings arg1) throws ScriptException {
+        return eval(arg0);
```

maybe stupid idea, but couldn't be `ctx().getBindings(id()).putMember()` be used to place binding values into Engine scope ?
Otherwise I'd suggest to log a warning if `!arg1.isEmpty()`.

---

It is not easy to support alternative bindings. Let's throw an exception: 26fbc8060f100